planner/core/plan_cost_detail.go

```diff
@@ -134,3 +176,177 @@ func setPhysicalIndexReaderCostDetail(p *PhysicalIndexReader, opt *physicalOptim
 		RowCountLbl, RowSizeLbl, NetworkFactorLbl, NetSeekCostLbl, ScanConcurrencyLbl))
 	opt.appendPlanCostDetail(detail)
 }
+
+func setPhysicalHashJoinCostDetail(p *PhysicalHashJoin, opt *physicalOptimizeOp, spill bool,
+	buildCnt, probeCnt, cpuFactor, rowSize, numPairs,
+	cpuCost, probeCPUCost, memCost, diskCost, probeDiskCost float64,
+	memQuota int64) {
+	if opt == nil {
+		return
+	}
+	detail := tracing.NewPhysicalPlanCostDetail(p.ID(), p.TP())
+	sessVars := p.ctx.GetSessionVars()
+	diskCostDetail := &HashJoinDiskCostDetail{
+		Spill:           spill,
+		UseOuterToBuild: p.UseOuterToBuild,
+		BuildRowCount:   buildCnt,
+		DiskFactor:      sessVars.GetDiskFactor(),
+		RowSize:         rowSize,
+		ProbeDiskCost: &HashJoinProbeDiskCostDetail{
+			SelectionFactor: SelectionFactor,
+			NumPairs:        numPairs,
+			HasConditions:   len(p.LeftConditions)+len(p.RightConditions) > 0,
+			Cost:            probeDiskCost,
+		},
+		Cost: diskCost,
+	}
+	memoryCostDetail := &HashJoinMemoryCostDetail{
+		Spill:         spill,
+		MemQuota:      memQuota,
+		RowSize:       rowSize,
+		BuildRowCount: buildCnt,
+		MemoryFactor:  sessVars.GetMemoryFactor(),
+		Cost:          memCost,
+	}
+	cpuCostDetail := &HashJoinCPUCostDetail{
+		BuildRowCount:     buildCnt,
+		CPUFactor:         cpuFactor,
+		ConcurrencyFactor: sessVars.GetConcurrencyFactor(),
+		ProbeCost: &HashJoinProbeCostDetail{
+			NumPairs:        numPairs,
+			HasConditions:   len(p.LeftConditions)+len(p.RightConditions) > 0,
+			SelectionFactor: SelectionFactor,
+			ProbeRowCount:   probeCnt,
+			Cost:            probeCPUCost,
+		},
+		HashJoinConcurrency: p.Concurrency,
+		Spill:               spill,
+		Cost:                cpuCost,
+	}
+
+	// record cpu cost detail
+	detail.AddParam(CPUCostDetailLbl, cpuCostDetail).
+		AddParam(CPUCostDescLbl, cpuCostDetail.desc()).
+		AddParam(ProbeCostDescLbl, cpuCostDetail.probeCostDesc())
+	// record memory cost detail
+	detail.AddParam(MemCostDetailLbl, memoryCostDetail).
+		AddParam(MemCostDescLbl, memoryCostDetail.desc())
+	// record disk cost detail
+	detail.AddParam(DiskCostDetailLbl, diskCostDetail).
+		AddParam(DiskCostDescLbl, diskCostDetail.desc()).
+		AddParam(ProbeDiskCostDescLbl, diskCostDetail.probeDesc())
+
+	detail.SetDesc(fmt.Sprintf("%s+%s+%s+all children cost", CPUCostDetailLbl, MemCostDetailLbl, DiskCostDetailLbl))
```

Comment:
Why don't we record the cost of children operators like previous PRs?

Reply:
Cost Detail is used to cost operator self cost, as all operators would be recorded cost and cost detail, thus we don't need to record child cost in cost detail.